File: po_executor.py
import re
from typing import Union
from playwright.async_api import async_playwright, TimeoutError as PWTimeout

PO_URL_LOGIN   = "https://pocketoption.com/en/login/"
PO_URL_CABINET = "https://pocketoption.com/en/cabinet/"
PO_URL_TRADE   = "https://pocketoption.com/en/cabinet/demo-quick-high-low/"

SEL = {
    "amount_box": "#put-call-buttons-chart-1 >> role=textbox",
    "expiry_5m_chip": ':text(":05:00")',
    "buy_link": 'a:has-text("Buy")',
    "sell_link": 'a:has-text("Sell")',
    "asset_open": ".pair-number-wrap",
    "assets_modal": "#modal-root .drop-down-modal-wrap.active",
    "assets_search": 'input[placeholder*="Search"]',
}

DEFAULT_TIMEOUT = 60_000

class PocketOptionExecutor:
    def __init__(self, email: str, password: str, headless: bool = False, storage_state: str = "po_storage.json"):
        self.email = email
        self.password = password
        self.headless = headless
        self.storage_state = storage_state
        self.p = None
        self.browser = None
        self.ctx = None
        self.page = None

    async def launch(self):
        self.p = await async_playwright().start()
        self.browser = await self.p.chromium.launch(headless=self.headless)
        try:
            self.ctx = await self.browser.new_context(
                storage_state=self.storage_state,
                user_agent=("Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) "
                            "AppleWebKit/537.36 (KHTML, like Gecko) "
                            "Chrome/121.0.0.0 Safari/537.36")
            )
        except Exception:
            self.ctx = await self.browser.new_context()
        self.page = await self.ctx.new_page()
        self.page.set_default_timeout(DEFAULT_TIMEOUT)

    async def close(self):
        try:
            if self.ctx:
                await self.ctx.storage_state(path=self.storage_state)
        except Exception:
            pass
        if self.browser:
            await self.browser.close()
        if self.p:
            await self.p.stop()

    async def login(self):
        await self.page.goto(PO_URL_CABINET, wait_until="domcontentloaded", timeout=DEFAULT_TIMEOUT)
        if "cabinet" in self.page.url:
            return
        await self.page.goto(PO_URL_LOGIN, wait_until="domcontentloaded", timeout=DEFAULT_TIMEOUT)
        try:
            await self.page.get_by_title("Close").click(timeout=3000)
        except Exception:
            pass
        await self.page.get_by_role("textbox", name="Email *").fill(self.email)
        await self.page.get_by_role("textbox", name="Password *").fill(self.password)
        await self.page.get_by_role("button", name="Sign In").click()
        print("\nSolve the CAPTCHA in the opened browser and finish login")

    async def goto_trade(self):
        await self.page.goto(PO_URL_TRADE, wait_until="domcontentloaded", timeout=DEFAULT_TIMEOUT)
        await self._dismiss_popups()  # ← added

    # --- NEW: minimal popup killer ---
    async def _dismiss_popups(self):
        """Close reward/news/cookie modals or toasts that can block clicks."""
        # try ESC once (many modals close on Escape)
        try:
            await self.page.keyboard.press("Escape")
            await self.page.wait_for_timeout(120)
        except Exception:
            pass

        selectors = [
            '#modal-root [data-test="close"]',
            '#modal-root [title="Close"]',
            '#modal-root button:has-text("Close")',
            '#modal-root button:has-text("OK")',
            '#modal-root button:has-text("Got it")',
            '.modal .icon-close, .modal .close, .popup .close',
            '.toast-close, .s-alert-close, [aria-label="Close"]',
        ]
        for sel in selectors:
            try:
                loc = self.page.locator(sel)
                if await loc.count():
                    await loc.first.click(timeout=600)
                    await self.page.wait_for_timeout(120)
            except Exception:
                continue

    async def _ensure_assets_modal_closed(self):
        try:
            await self.page.wait_for_selector(SEL["assets_modal"], state="hidden", timeout=1000)
        except Exception:
            try:
                await self.page.keyboard.press("Escape")
                await self.page.wait_for_selector(SEL["assets_modal"], state="hidden", timeout=2000)
            except Exception:
                pass

    async def _open_closed_trades(self):
        try:
            await self.page.click('text=Trades', timeout=4000)
        except Exception:
            pass
        try:
            await self.page.get_by_text("Closed", exact=True).click(timeout=3000)
        except Exception:
            try:
                await self.page.get_by_role("tab", name="Closed").click(timeout=3000)
            except Exception:
                pass
        await self.page.wait_for_timeout(700)

    async def select_pair(self, pair: str, max_retries: int = 3):
        """
        Type the base pair only (e.g., 'EUR/JPY' or 'EURJPY'), then click the '… OTC' row.
        Verifies header shows '<BASE> OTC'. Retries up to max_retries, then raises.
        """
        base = pair.upper().replace(" OTC", "").strip()
        expected = f"{base} OTC"
        keyword = base.replace("/", "")

        for attempt in range(1, max_retries + 1):
            await self.page.click(SEL["asset_open"], timeout=DEFAULT_TIMEOUT)
            modal = self.page.locator(SEL["assets_modal"])
            await modal.wait_for(state="visible", timeout=5000)

            try:
                search = modal.get_by_placeholder("Search")
            except Exception:
                search = modal.locator(SEL["assets_search"])

            await search.click()
            try:
                await self.page.keyboard.press("Meta+A")
                await self.page.keyboard.press("Backspace")
            except Exception:
                pass
            await search.fill(keyword)
            await self.page.wait_for_timeout(350)

            clicked = False
            try:
                await modal.get_by_text(expected, exact=False).first.click(timeout=1200)
                clicked = True
            except Exception:
                try:
                    await modal.get_by_text(base, exact=False).first.click(timeout=1200)
                    clicked = True
                except Exception:
                    clicked = False

            try:
                await self.page.wait_for_selector(SEL["assets_modal"], state="hidden", timeout=1500)
            except Exception:
                try:
                    await self.page.keyboard.press("Enter")
                except Exception:
                    pass
                await self.page.keyboard.press("Escape")
                await self.page.wait_for_selector(SEL["assets_modal"], state="hidden", timeout=2500)

            await self.page.wait_for_timeout(400)

            active = (await self.page.locator(SEL["asset_open"]).inner_text()).strip().upper()
            if clicked and base in active and "OTC" in active:
                return
            print(f"[WARN] Active pair '{active}' != '{expected}' (try {attempt}/{max_retries})")

        raise PWTimeout(f"Could not select pair '{expected}' after {max_retries} attempts")

    # -------- FIXED: robust amount setter (Mac-friendly & verified) --------
    async def set_amount(self, amount: Union[float, int]):
        await self._ensure_assets_modal_closed()
        amt = self.page.locator(SEL["amount_box"])
        await amt.wait_for(state="visible", timeout=5000)

        want = f"{float(amount):.2f}".rstrip("0").rstrip(".")

        for _ in range(3):
            try:
                await amt.fill("")          # hard clear
                await amt.fill(want)        # set
                await self.page.wait_for_timeout(120)

                cur = await amt.input_value()
                if cur == want or cur == f"{float(amount):.2f}":
                    return
            except Exception:
                pass
            # fallback: try select-all on both Mac/Win
            try:
                await amt.click()
                for combo in ("Meta+A", "Control+A"):
                    try:
                        await self.page.keyboard.press(combo)
                        await self.page.keyboard.press("Backspace")
                    except Exception:
                        pass
                await amt.fill(want)
            except Exception:
                pass

        cur = await amt.input_value()
        raise RuntimeError(f"Amount set failed: wanted {want}, got {cur}")

    async def set_expiry(self, minutes: int):
        if minutes == 5:
            try:
                await self.page.locator(SEL["expiry_5m_chip"]).click(timeout=3000)
                return
            except Exception:
                pass
        try:
            await self.page.click('[data-test="expiry-dropdown"]', timeout=3000)
            if minutes == 5:
                try:
                    await self.page.click('text="5 min"', timeout=2000)
                    return
                except Exception:
                    return
            else:
                await self.page.click(f'text="{minutes} min"', timeout=3000)
        except Exception:
            if minutes != 5:
                raise

    async def click_direction(self, direction: str):
        d = direction.upper()
        try:
            if d == "BUY":
                await self.page.locator(SEL["buy_link"]).first.click(timeout=3000)
                return
            elif d == "SELL":
                await self.page.locator(SEL["sell_link"]).first.click(timeout=3000)
                return
            else:
                raise ValueError(f"Unknown direction {direction}")
        except Exception:
            pass
        try:
            await self.page.get_by_text(d, exact=False).first.click(timeout=3000)
            return
        except Exception as e:
            raise PWTimeout(f"Could not click {d} button") from e

    async def place_trade(self, pair: str, direction: str, expiry_min: int, amount: Union[float, int]):
        await self.goto_trade()
        await self._dismiss_popups()  # ← added
        await self.select_pair(pair)
        await self.set_amount(amount)
        await self.set_expiry(expiry_min)
        await self._dismiss_popups()  # ← added
        await self.click_direction(direction)

    async def last_closed_profit(self) -> float:
        await self._open_closed_trades()
        row = self.page.locator('.deals-list__item-first, [data-test="trades-closed"] [class*="row"], .table-row').first
        txt = await row.inner_text(timeout=3000)
        amounts = re.findall(r'([+\-]?\s*\$?\s*\d+(?:[\.,]\d{1,2})?)', txt)
        if not amounts:
            raise RuntimeError(f"No profit figure found in closed row text: {txt!r}")
        raw = amounts[-1]
        raw = (raw.replace(" ", "")
               .replace("$", "")
               .replace(",", "")
               .replace("\u202f", "")
               .replace("\xa0", ""))
        try:
            return float(raw)
        except ValueError:
            norm = raw.replace("+", "").replace("-", "")
            return float(norm) if norm else 0.0

    async def check_last_closed_won(self) -> bool:
        profit = await self.last_closed_profit()
        return profit > 0.0

File: po_executor.js
// po_executor.js — Executor with robust Closed tab scraping for /peek
// Hardened against stale results, formatting mismatches, and cross-chain contamination.

const path = require("path");
const express = require("express");
const { chromium } = require("playwright");
const fs = require("fs");
require("dotenv").config();

const PO_URL_TRADE = "https://pocketoption.com/en/cabinet/";
const HEADLESS = process.env.HEADLESS === "1";
const DEFAULT_TIMEOUT = 60_000;
const LOG_FILE = path.resolve(__dirname, "trade_log.csv");

const SCREEN_DIR = path.resolve(__dirname, "screens");
if (!fs.existsSync(SCREEN_DIR)) fs.mkdirSync(SCREEN_DIR);

const SEL = {
  symbolToggle: 'span.current-symbol.current-symbol_cropped, .current-symbol',
  assetOverlay: '.drop-down-modal-wrap.active',
  tradePanel: '[id^="put-call-buttons-chart"]',
  searchInput: 'input[placeholder="Search"]',

  buyBtn: '#put-call-buttons-chart-1 a.buy, #put-call-buttons-chart-1 button:has-text("Buy"), a.btn.btn-call',
  sellBtn: '#put-call-buttons-chart-1 a.sell, #put-call-buttons-chart-1 button:has-text("Sell"), a.btn.btn-put',

  closedTab: 'li:has-text("Closed")',
  closedRow: '.deals-list__item',
  directionUp: 'i.fa.fa-arrow-up',
  directionDown: 'i.fa.fa-arrow-down',
  profitCell: '.centered',
  closeTime: '.close-time'
};

let context, page;
let tradeInProgress = false;
let recentTrades = {};
let lastPairCache = null;

// remembers the most recent placed trade params per chain
const expectedByChain = Object.create(null);

const sleep = (ms) => new Promise(r => setTimeout(r, ms));

async function withRetry(fn, attempts = 2, label = "op") {
  let lastErr;
  for (let i = 0; i < attempts; i++) {
    try { return await fn(); }
    catch (err) {
      lastErr = err;
      console.warn(`[Retry] ${label} failed (${i + 1}/${attempts}) -> ${err?.message}`);
      await sleep(200);
    }
  }
  throw lastErr;
}

async function waitForTradePanel() {
  await page.waitForSelector(SEL.tradePanel, { timeout: DEFAULT_TIMEOUT });
  console.log("[✅] Trade panel detected");
}

async function forceCloseOverlays() {
  try {
    const overlayVisible = await page.locator("div.mfp-wrap, .drop-down-modal-wrap.active").first().isVisible().catch(() => false);
    if (!overlayVisible) return;

    await page.evaluate(() => {
      document.querySelectorAll("div.mfp-bg, div.mfp-wrap, .drop-down-modal-wrap.active").forEach(el => el.remove());
    }).catch(() => {});

    await sleep(400);
    const stillVisible = await page.locator("div.mfp-wrap, .drop-down-modal-wrap.active").first().isVisible().catch(() => false);
    if (stillVisible) {
      console.log("[Heal] Overlay stuck, reloading page");
      await page.reload({ waitUntil: "domcontentloaded", timeout: 15000 }).catch(() => {});
      await ensureOnPO();
    }
  } catch (err) {
    console.error("[Overlay] Force close failed:", err.message);
  }
}

async function ensureOnPO() {
  if (!page || page.isClosed()) throw new Error("No page");
  const url = page.url() || "";
  if (!url.includes("pocketoption.com")) {
    console.log("[Nav] Navigating to PocketOption trade page");
    await page.goto(PO_URL_TRADE, { waitUntil: "domcontentloaded", timeout: DEFAULT_TIMEOUT });
  }
  await withRetry(async () => { await waitForTradePanel(); }, 2, "wait trade panel");
}

async function ensurePageAlive() {
  if (!page || page.isClosed()) {
    console.log("[Heal] Page closed. Restarting browser");
    await initBrowser();
    return;
  }
  const panelVisible = await page.locator(SEL.tradePanel).first().isVisible().catch(() => false);
  if (!panelVisible) {
    console.log("[Heal] Trade panel not visible. Reloading");
    await page.reload({ waitUntil: "domcontentloaded", timeout: DEFAULT_TIMEOUT }).catch(() => {});
    await ensureOnPO();
  }
}

async function setTradeAmount(amount) {
  const panel = page.locator(SEL.tradePanel).first();
  const amountBox = panel.getByRole('textbox').first();
  await amountBox.waitFor({ state: 'attached', timeout: DEFAULT_TIMEOUT }).catch(() => {});
  try {
    await amountBox.fill(String(amount), { force: true, timeout: 1500 });
    return;
  } catch {}
  await panel.click({ force: true }).catch(() => {});
  await page.keyboard.press('Control+A').catch(() => {});
  await page.keyboard.press('Backspace').catch(() => {});
  await page.keyboard.type(String(amount)).catch(() => {});
}

async function selectPair(pair) {
  if (lastPairCache && lastPairCache.toLowerCase() === pair.toLowerCase()) {
    console.log(`[Cache] Skipping selectPair, reusing: ${pair}`);
    return;
  }

  const toggle = page.locator(SEL.symbolToggle).first();
  try {
    await withRetry(async () => {
      await toggle.click({ timeout: DEFAULT_TIMEOUT });
      await page.waitForSelector(SEL.assetOverlay, { state: 'visible', timeout: DEFAULT_TIMEOUT });
    }, 2, "open asset overlay");

    const cleaned = pair.replace(" OTC", "").replace("/", "").toLowerCase();
    const search = page.locator(SEL.searchInput).first();
    await search.fill("");
    await search.type(cleaned, { delay: 30 }).catch(() => {});
    await sleep(200);

    const listItem = page.locator('.alist__label', { hasText: pair }).first();
    await withRetry(async () => { await listItem.click({ timeout: DEFAULT_TIMEOUT }); }, 2, "select list item");

    console.log(`[Step] Selected pair: ${pair}`);
    await page.keyboard.press('Escape').catch(() => {});
    await forceCloseOverlays();
    await sleep(100);

    lastPairCache = pair;
  } catch (err) {
    const ts = Date.now();
    const screenshotPath = path.join(SCREEN_DIR, `selectPair_fail_${pair}_${ts}.png`);
    try {
      await page.screenshot({ path: screenshotPath, fullPage: true });
      console.log(`[📸] Saved screenshot on selectPair fail: ${screenshotPath}`);
    } catch (ssErr) {
      console.error("[❌] Screenshot failed:", ssErr.message);
    }
    throw err;
  }
}

function appendLog(ts, pair, dir, amount, result, profit, ml_tag = "", chain_id = "", closed_at = "") {
  const header = "Time,Pair,Dir,Amount,Result,Profit,ML_Tag,Chain_ID,Closed_At\n";
  if (!fs.existsSync(LOG_FILE)) {
    fs.writeFileSync(LOG_FILE, header);
  }
  fs.appendFileSync(LOG_FILE, `${ts},${pair},${dir},${amount},${result},${profit},${ml_tag},${chain_id},${closed_at}\n`);
}

function recordClosedTrade(meta) {
  const { chain_id } = meta;
  if (!chain_id) return;
  if (!recentTrades[chain_id]) recentTrades[chain_id] = [];
  recentTrades[chain_id].unshift(meta);
  recentTrades[chain_id] = recentTrades[chain_id].filter(
    t => Date.now() - new Date(t.closed_at).getTime() < 600000
  );
  if (recentTrades[chain_id].length > 5) recentTrades[chain_id].pop();
}

async function parseClosedTrade(amount, pair, direction, ml_tag, chain_id = "") {
  let profit = 0.0, result = "LOSS", closed_at = new Date().toISOString();
  try {
    await page.locator(SEL.closedTab).click({ timeout: 2000 });
    await page.waitForSelector(SEL.closedRow, { timeout: 3000 });
    await page.waitForTimeout(500);

    const rows = await page.locator(SEL.closedRow).all();
    for (const row of rows) {
      const text = await row.innerText().catch(() => '');

      // --- Pair check
      if (pair && !text.includes(pair)) continue;

      // --- Amount check (numeric tolerant)
      const numericMatches = text.match(/[\d.,]+/g) || [];
      const hasAmount = numericMatches.some(num => {
        const val = parseFloat(num.replace(/,/g, ""));
        return !isNaN(val) && Math.abs(val - amount) < 0.01;
      });
      if (amount && !hasAmount) continue;

      // --- Direction check
      let detectedDirection = null;
      if (await row.locator(SEL.directionUp).count() > 0) detectedDirection = "BUY";
      if (await row.locator(SEL.directionDown).count() > 0) detectedDirection = "SELL";
      if (detectedDirection && direction && detectedDirection !== direction.toUpperCase()) continue;

      // --- Profit check
      const profitNode = row.locator(SEL.profitCell).last();
      if (await profitNode.count() === 0) continue;
      let profitText = await profitNode.innerText();
      profitText = profitText.replace(/[^\d.-]/g, "");
      profit = parseFloat(profitText || "0");
      result = profit > 0 ? "WIN" : "LOSS";

      // --- Closed time freshness
      try {
        const timeNode = row.locator(SEL.closeTime).first();
        if (await timeNode.count()) {
          const timeStr = await timeNode.innerText();
          const now = new Date();
          closed_at = new Date(now.toDateString() + " " + timeStr + " UTC").toISOString();
        }
      } catch {}
      const closedAgo = Date.now() - new Date(closed_at).getTime();
      if (closedAgo > 180000) continue; // skip if older than 3 min

      break;
    }
  } catch (err) {
    console.error("[❌] Result parse failed:", err.message);
  }

  const meta = { amount, pair, direction, ml_tag, chain_id, profit, result, closed_at };
  recordClosedTrade(meta);

  const ts2 = new Date().toISOString();
  appendLog(ts2, pair, direction, amount, result, profit, ml_tag, chain_id, closed_at);
  console.log(`[Result] ${result} ${pair} ${direction} $${amount} profit=${profit} [${ml_tag}] [chain=${chain_id}] closed_at=${closed_at}`);
  return meta;
}

async function placeTrade(pair, amount, direction, ml_tag = "", chain_id = "", expiration = 300) {
  if (tradeInProgress) {
    console.warn("[Guard] Trade already in progress. Skipping duplicate request.");
    return { success: false, result: "SKIPPED", profit: 0, ml_tag, chain_id };
  }
  tradeInProgress = true;

  try {
    console.log(`[Step] Trade request: ${direction.toUpperCase()} ${pair} $${amount} [${ml_tag}] [chain=${chain_id}]`);
    await ensurePageAlive();
    await ensureOnPO();
    await forceCloseOverlays();

    await withRetry(async () => { await selectPair(pair); }, 2, "selectPair");
    await withRetry(async () => { await setTradeAmount(amount); }, 2, "setTradeAmount");

    const panel = page.locator(SEL.tradePanel).first();
    const btn = direction.toLowerCase() === 'buy'
      ? panel.locator(SEL.buyBtn).first()
      : panel.locator(SEL.sellBtn).first();

    await btn.scrollIntoViewIfNeeded().catch(() => {});
    await btn.waitFor({ state: 'visible', timeout: DEFAULT_TIMEOUT });
    console.log(`[CLICK] ${direction.toUpperCase()} button for ${pair} @ $${amount}`);

    await btn.click({ timeout: DEFAULT_TIMEOUT, force: true }).catch(err => {
      console.error("[❌] Button click failed:", err.message);
    });

    if (chain_id) {
      expectedByChain[chain_id] = { amount, pair, direction, ml_tag, ts: Date.now() };
    }

    const ts = new Date().toISOString();
    appendLog(ts, pair, direction, amount, "OPEN", 0.0, ml_tag, chain_id, "");

    return { success: true, result: "OPEN", profit: 0, ml_tag, chain_id };
  } finally {
    tradeInProgress = false;
  }
}

async function peekLatestProfit(chain_id = null) {
  await ensurePageAlive();
  try { await page.locator(SEL.closedTab).click({ timeout: 3000 }); } catch {}

  if (!chain_id) return null;
  const expected = expectedByChain[chain_id];
  if (!expected) return null;

  const { amount, pair, direction, ml_tag } = expected;
  const meta = await parseClosedTrade(amount, pair, direction, ml_tag, chain_id);
  return meta || null;
}

const app = express();
app.use(express.json());

app.post("/trade", (req, res) => {
  console.log("[REQ] Incoming trade request:", req.body);
  const { pair, amount, direction, ml_tag, chain_id, expiration } = req.body || {};
  if (!pair || !amount || !direction) {
    return res.status(400).json({ success: false, error: "pair, amount, direction required" });
  }

  res.json({ success: true, result: "QUEUED", pair, amount, direction, ml_tag, chain_id });

  (async () => {
    try { await placeTrade(pair, amount, direction, ml_tag, chain_id, expiration); }
    catch (err) { console.error("[❌] Background trade failed:", err); }
  })();
});

app.get("/peek", async (req, res) => {
  try {
    const { chain_id } = req.query;
    const result = await peekLatestProfit(chain_id);
    if (!result) return res.json({ ok: false, profit: 0, ml_tag: "", chain_id: "", closed_at: null });
    return res.json({ ok: true, profit: result.profit, ml_tag: result.ml_tag, chain_id: result.chain_id, closed_at: result.closed_at });
  } catch (err) {
    return res.json({ ok: false, error: err?.message || String(err), profit: 0, ml_tag: "", chain_id: "", closed_at: null });
  }
});

async function initBrowser() {
  console.log("[Init] Launching PocketOption");
  const browser = await chromium.launch({ headless: HEADLESS, args: ["--no-sandbox", "--disable-dev-shm-usage"] });
  context = await browser.newContext({ storageState: "po_storage.json" });
  page = await context.newPage();
  page.setDefaultTimeout(DEFAULT_TIMEOUT);

  await page.goto(PO_URL_TRADE, { waitUntil: "domcontentloaded", timeout: DEFAULT_TIMEOUT });
  await ensureOnPO();
  console.log("[Init] PocketOption ready");
}

app.listen(3000, async () => {
  await initBrowser();
  console.log("[Server] Executor API listening on http://localhost:3000");
});

process.on("SIGINT", async () => { try { await context?.close(); } catch {} process.exit(0); });
process.on("unhandledRejection", (err) => console.error("[UnhandledRejection]", err));
process.on("uncaughtException", (err) => console.error("[UncaughtException]", err));
